### PR TITLE
Add separate commands for create/fetch/exercise by template/interface.

### LINKED ISSUE #11674: Make validation replay by interface if by_interface is set
unsafeTranslateTransactionRoots should call by interface if `by_interface` is set. Current code however should already work.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
Resolves #11674 and #11675

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by sofiafaro-da** [comment]
I'm going to add engine tests that use these commands.

I'm not going to add anything in the CommandPreprocessorSpec yet though, since it requires interfaces in the LF parser. I'm adding that to issue https://github.com/digital-asset/daml/issues/10917

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala [hunk 1]
```diff
@@ -15,30 +15,79 @@ sealed abstract class Command extends Product with Serializable {
...
+      interfaceId: Identifier,
+      contractId: SContractId,
+      choiceId: ChoiceName,
+      argument: SValue,
+  ) extends Command {
+    // TODO https://github.com/digital-asset/daml/issues/11342
+    //   The actual template id isn't known until run time.
+    //   The interface id is the best we've got.
+    val templateId = interfaceId
+  }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:59]:
```suggestion
  ) extends Command 
```
This seems unused. Better to remove it from `Command`

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala [hunk 2]
```diff
@@ -41,6 +48,36 @@ final case class ExerciseCommand(
...
+    interfaceId: Identifier,
+    contractId: Value.ContractId,
+    choiceId: ChoiceName,
+    argument: Value,
+) extends Command {
+  // TODO https://github.com/digital-asset/daml/issues/11342
+  //   The actual template id isn't known until run time.
+  //   The interface id is the best we've got.
+  val templateId = interfaceId
+}
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala:79]: Same think, Command#template is not used. Better to delete it.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala [hunk 3]
```diff
@@ -15,30 +15,79 @@ sealed abstract class Command extends Product with Serializable {
...
+  /** Exercise a template choice, by interface */
+  final case class ExerciseByInterface(
+      interfaceId: Identifier,
+      templateId: Identifier,
+      contractId: SContractId,
+      choiceId: ChoiceName,
+      argument: SValue,
+  ) extends Command
+
+  /** Exercise an interface choice */
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:48]:
```suggestion
  /** Exercise an interface choice by template*/
```
maybe? if not I think I’m confused

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala [hunk 4]
```diff
@@ -70,11 +107,35 @@ final case class CreateAndExerciseCommand(
     choiceArgument: Value,
 ) extends ApiCommand
 
+/** Fetch either a template or interface */
 final case class FetchCommand(
     templateId: Identifier,
     coid: Value.ContractId,
 ) extends Command
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala:114]: No need for this one. 
This is used only by canton for replay. They have the info need. 

Only the `APICommand` needs to be generic.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala [hunk 5]
```diff
@@ -15,30 +15,79 @@ sealed abstract class Command extends Product with Serializable {
...
+  final case class ExerciseInterface(
+      interfaceId: Identifier,
+      contractId: SContractId,
+      choiceId: ChoiceName,
+      argument: SValue,
+  ) extends Command {
+    // TODO https://github.com/digital-asset/daml/issues/11342
+    //   The actual template id isn't known until run time.
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:56]: I don’t understand this. This is not an `APICommand` at least judging from the fact that it doesn’t extend that. So it should only be used for replays and there we should know the  template id.

What am I missing?

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 6]
```diff
@@ -33,34 +33,100 @@ private[lf] final class CommandPreprocessor(
       templateId: Ref.Identifier,
       argument: Value,
   ): speedy.Command.Create = {
+    discard[Ast.TemplateSignature](handleLookup(interface.lookupTemplate(templateId)))
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:36]: :+1: 
Indeed the processor should validate thoughtfully the command.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 7]
```diff
@@ -134,12 +214,32 @@ private[lf] final class CommandPreprocessor(
...
-        //  -- handle the case where templateId is an interface
+      case command.FetchCommand(identifier, coid) =>
+        val templateOrInterface = handleLookup(interface.lookupTemplateOrInterface(identifier))
+        val cid = valueTranslator.unsafeTranslateCid(coid)
+        templateOrInterface match {
+          case Left(template @ _) =>
+            speedy.Command.Fetch(identifier, cid)
+          case Right(interface @ _) =>
+            speedy.Command.FetchInterface(identifier, cid)
+        }
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:225]: We do not need the generic case for Fetch.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 8]
```diff
@@ -134,12 +214,32 @@ private[lf] final class CommandPreprocessor(
...
+      case command.FetchTemplateCommand(templateId, coid) => {
         discard[Ast.TemplateSignature](handleLookup(interface.lookupTemplate(templateId)))
```
Comment (remyhaemmerle-da) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:227]: I am not sure the type parameter is very usefull in those `discard`

**Review by remyhaemmerle-da — APPROVED** [review]
Nice.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala [hunk 9]
```diff
@@ -70,11 +107,35 @@ final case class CreateAndExerciseCommand(
     choiceArgument: Value,
 ) extends ApiCommand
 
+/** Fetch either a template or interface */
```
Comment (cocreature) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala:110]: why do we need this? `fetch` is not a ledger API command so afaict we should never need this.

**Review by cocreature — APPROVED** [review]
thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:56]: :+1:

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala (hunk 7 above)
Comment (sofiafaro-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:225]: 👍  I'll remove it then, thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 5 above)
Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:56]: You're missing nothing, I didn't realize it wasn't needed. I will remove it.

Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:56]: This Speedy command is generated by the commandPreprocessor when invoking `ExerciseCommand` on an interface id directly.

Speedy commands should already know if they are invoking an interface id (generically, like this) or a template id.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 1 above)
Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:59]: It's used in the engine. https://github.com/digital-asset/daml/blob/main/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Engine.scala#L120

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 3 above)
Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:48]: Not by template. It exercises an interface without knowing the template id. This is what you get when exercising an interface via the ledger api, e.g.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala (hunk 4 above)
Comment (sofiafaro-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala:114]: Should I remove `FetchInterface` below as well? Doesn't sound useful by the same reason.

Comment (sofiafaro-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala:114]: And `ExerciseInterfaceCommand` as well.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 3 above)
Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:48]: Ah I see, maybe make it explicit that this is used via the ledger API.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala:59]: I missed it. I think it is OK for now.

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala (hunk 8 above)
Comment (sofiafaro-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala]: Dropped it, thanks!

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala (hunk 4 above)
Comment (sofiafaro-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala]: I dropped them.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala (hunk 3 above)
Comment (sofiafaro-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Command.scala]: Added a comment.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala (hunk 9 above)
Comment (sofiafaro-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/command/Command.scala]: Dropped it

**Comment by sofiafaro-da** [comment]
Enabling auto-merge so I can work on engine tests in a separate PR.

**Pull request merged**